### DISCUSSION TITLE
How to load `rubygems_plugin.rb` when running `bundle install`

### DISCUSSION THREAD

**Discussion by impostorsguides** [rbenv/rbenv#1516] [Q&A] [answered]
I'm reading the RBENV codebase, and I came upon its `rubygems_plugin.rb` file.  It includes the following code:

```
if defined?(Bundler::Installer) && Bundler::Installer.respond_to?(:install) && !Bundler::Installer.respond_to?(:install_without_rbenv_rehash)
  Bundler::Installer.class_eval do
  ...
end
```

According to [the PR that introduced it](https://github.com/rbenv/rbenv/pull/806/files), the goal of this code is to ensure that RBENV rebuilds its shims only at the end of the `bundle install` process, as opposed to after each gem installation.  To me, this implies that `bundle install` will run the `rubygems_plugin.rb` file.

I wanted to step through the code to see how it works, so I did the following:

 - added debugger statements to the `rubygems_plugin.rb` file inside my `~/.rbenv` directory (both just before, as well as inside, the above `if` block)
 - installed a fresh Ruby version (`v3.1.4`) via `rbenv install 3.1.4` to make sure that any gems in a Gemfile would not yet be installed
 - created a fresh Rails project (`rails new foobar`) using that new Ruby version, so that I could run `bundle install` on its Gemfile
 - Added a gem to the Gemfile (`wisper`) which I knew contained an executable, and would therefore return `true` for [this line of code](https://github.com/rbenv/rbenv/blob/master/rbenv.d/exec/gem-rehash/rubygems_plugin.rb#L4) and result in a shim being generated.
 - ran `bundle install` inside that project.

I was surprised to see that I didn't hit my debuggers.  I also checked `~/.rbenv/shims` for the `wisper` shim, but I didn't find it. 
 Both of these things imply that the file isn't being loaded during a `bundle install`.  However, when I run a simple `gem install wisper`, I do hit the debugger outside the `if` block, and I do see the `wisper` shim.

I also tried to install the version of Ruby which was active at the time of the original PR (`v2.1.2`), but it's no longer supported (and in fact, can no longer be installed on my 2023 Macbook).

I read through [the Bundler CHANGELOG.md file](https://github.com/rubygems/bundler/blob/master/CHANGELOG.md) from the deprecated Bundler repo, to see if maybe support for this file was removed from the `bundle install` command.  Nothing stood out to me, but maybe I missed something.

I'm guessing this is user error, and that I'm simply missing something.  But I thought I'd post this question anyway, in case this file no longer loaded when `bundle install` is run?

Also FYI: I also read through the Bundler git history, and I saw [this PR](https://github.com/rubygems/rubygems/commit/cb5c3e17774fb8fcc513348bc5e06e1b3f21cc6b) from June 2018 that introduced a `GEM_AFTER_INSTALL_ALL` hook to Bundler.  This appears to address the same goal that [the above RBENV PR from 2015](https://github.com/rbenv/rbenv/pull/806/files) addressed.  Thought I'd call that out in case it's useful.

**Comment by mislav** [comment, marked as answer]
Thanks for the detailed dive! I didn't know this, and you're right: in my testing it looks like `bundle` commands no longer cause the `rubygems_plugin.rb` file to be loaded from LOAD_PATH.

You are also right that Bundler now has its own plugin API, but it doesn't look like rbenv can automatically inject its own plugin via environment variables. Instead, it looks like a user needs to explicitly enable a plugin via a `bundle` command.

I'd be fine with giving up on rbenv hooking into `bundle install`, and instead just leaving the `gem install` hook. People who use bundler can easily generate binstubs for their own project, which alleviates the need for rbenv to keep its shims up to date.

**Reply by impostorsguides** [reply]
Thanks for the clarification.  If the Bundler-specific code is no longer executed, is it worthwhile to restore `rubygems_plugin.rb` to its pre-Bundler state?  If so, I'd be happy to put up a PR changing it to something like this:

```
hook = lambda do |installer|
  begin
    # Ignore gems that aren't installed in locations that rbenv searches for binstubs
    if installer.spec.executables.any? &&
        [Gem.default_bindir, Gem.bindir(Gem.user_dir)].include?(installer.bin_dir)
      `rbenv rehash`
    end
  rescue
    warn "rbenv: error in gem-rehash (#{$!.class.name}: #{$!.message})"
  end
end

begin
  Gem.post_install(&hook)
  Gem.post_uninstall(&hook)
rescue
  warn "rbenv: error installing gem-rehash hooks (#{$!.class.name}: #{$!.message})"
end
```

**Comment by impostorsguides** [comment]
If you're as curious as I was:

Upon further research, it looks like this unexpected behavior was caused by [this PR](https://github.com/rubygems/bundler/pull/4954/files) in the Bundler repo, specifically the removal of this line in `lib/bundler/cli/install.rb`:

```
Gem.load_env_plugins if Gem.respond_to?(:load_env_plugins)
```

When I edit this file on my machine to add this line back and re-run `bundle install` in my toy Rails project, I once again hit the debugger statements in RBENV's `rubygems_plugin.rb` file.

This change was, in fact, mentioned in the Bundler CHANGELOG.md file, I simply glossed over it the first time around:

> load all rubygems plugins when installing gems (https://github.com/rubygems/bundler/issues/2824, @segiddins)

I found the Bundler released which introduced this behavior by installing every Bundler version from the last 8 years (not that many, really), and running `bundle _version_number_ install` (i.e. `bundle _1.14.0_ install`) on each one until I started hitting my breakpoints again, then combing through the release notes for that version.

I've already got a question open on the RubyGems repo [here](https://github.com/rubygems/rubygems/discussions/6812), and I'll add the above remarks to the question and see if they think it's worth a PR.

**Comment by impostorsguides** [comment]
I was curious how one would go about hooking into Bundler to run `rbenv rehash`, so I took the liberty of creating [this](https://github.com/richiethomas/rbenv_bundler_rehash) Ruby gem as a very rough proof-of-concept.  No worries if you'd rather not go down this route- it was still a useful learning exercise for me.

TL;DR- running `gem install rbenv_bundler_rehash` followed by `bundle install plugin rbenv_bundler_rehash` will result in gems being shimmed whenever `bundle install` is run.

I'm having credential-related problems while trying to run `rake release` in order to push this gem up to RubyGems, so for now `gem install rbenv_bundler_rehash` won't work.  But feel free to inspect the code.

Lastly, it turns out that there is no "AFTER_UNINSTALL_ALL" lifecycle hook in Bundler (presumably because there's no `bundle uninstall` command), so there's no way to add this logic to uninstalling gems.
